docs/latest/design/ratelimit.md

```diff
@@ -0,0 +1,66 @@
+# Ratelimiting Design
+
+## Overview
+
+Ratelimiting is a feature that allows the user to limit the number of incoming requests
+to a predefined value based on attributes within the traffic flow.
+
+Here are some reasons why a user may want to implements Ratelimits
+
+* To prevent malicious activity such as DDoS attacks.
+* To prevent applications and its resources (such as a database) from getting overloaded.
+* To create API limits based on user entitlements.
+
+## API
+
+* Here is an example of a ratelimit implemented by the platform engineer that limits requests made
+by every unique client remote address to 100 requests per second, to help mitigate DDoS attacks.
+```
+apiVersion: gateway.envoyproxy.io/v1alpha1
+kind: RateLimiting 
+metadata:
+  name: ratelimit-per-client-ip
+spec:
+  type: Global
+  rules:
+  - matches:
+    - remoteAddress: {}
+    limit:
+      requests: 100
+      unit: Minute
+```
+
+* Here is an example of a ratelimit implemented by the application developer that limits total requests made
+to a specific route to safeguard health of internal application components.
+```
+apiVersion: gateway.envoyproxy.io/v1alpha1
+kind: RateLimiting
+metadata:
+  name: ratelimit-all-requests
+spec:
+  type: Global
+  rules:
+  - matches:
+    - limit:
+        requests: 1000
+	unit: Second
+```
+
+* Here is an example of a ratelimit implemented by the application developer to limit a specific set of users
+by matching on a custom `x-user-tier` header with a value set to `one`
+```
+apiVersion: gateway.envoyproxy.io/v1alpha1
+kind: RateLimiting
+metadata:
+  name: ratelimit-specific-requests
+spec:
+  type: Global
+  rules:
+  - matches:
```

Comment:
`matches` is unneeded if RateLimit is applied to an HTTPRoute. For example:

```yaml
apiVersion: gateway.networking.k8s.io/v1beta1
kind: HTTPRoute
metadata:
  name: example
spec:
  ...
  rules:
    - matches:
        headers:
          - name: x-user-tier
            value: one
      filters:
        - type: ExtensionRef
          extensionRef:
            group: gateway.envoyproxy.io
            kind: RateLimiting
            name: ratelimit-specific-requests
      backendRefs:
        - name: httpbin
          port: 80
```

Matching criteria can be derived from the HTTPRoute that references the RateLimiting filter and RateLimiting can define the rate limit policy, e.g. [domain and descriptors](https://github.com/envoyproxy/ratelimit#definitions) (for type `Global`).

Reply:
the `matches` criteria within HTTPRoute helps map a specific traffic flow to a destination backend.
Even though the `RateLimit` is linked to a specific `HTTPRoute`, the `matches` criteria here further maps the traffic to specific users (source ip, header etc)

Reply:
It feels to me like this is the biggest decision here - adding an extra layer of matching inside the ratelimit filter rather than having it filter whatever is configured in the HTTPRoute is an interesting approach - it's not really what I expected when thinking about this, tbh.

@arkodg, do you think you could put some explanation in this doc (maybe in the decisions made section) as to why you've chosen this way rather than the one Daneyon suggested? I'm not saying we should change it, just that we should record the reasons we've decided this way.

Reply:
agreed, ive added a section on it under `Design Decisions`